File: src/core/executer/validation.py
from typing import Sequence, Dict, Optional
from pathlib import Path

from .exceptions import CommandValidationError

def validate_command(cmd: Sequence[str]) -> None:
    """
    Validate command sequence
    
    Args:
        cmd: Command sequence as strings
        
    Raises:
        CommandValidationError: If command is invalid
    """
    if not cmd:
        raise CommandValidationError("Empty command sequence")
    
    if not isinstance(cmd, Sequence):
        raise CommandValidationError(
            f"Command must be a sequence, got {type(cmd)}"
        )

    for arg in cmd:
        if not isinstance(arg, str):
            raise CommandValidationError(
                f"Command arguments must be strings, got {type(arg)}"
            )
        
        # Basic injection prevention
        if any(char in arg for char in ['&', '|', ';', '`', '$', '\\']):
            raise CommandValidationError(
                f"Invalid command argument: {arg}"
            )

def validate_env(env: Optional[Dict[str, str]]) -> Dict[str, str]:
    """
    Validate and sanitize environment variables
    
    Args:
        env: Environment variables dictionary
        
    Returns:
        Sanitized environment dictionary
    """
    if not env:
        return {}
        
    sanitized = {}
    for key, value in env.items():
        if not isinstance(key, str) or not isinstance(value, str):
            raise CommandValidationError(
                "Environment variables must be strings"
            )
            
        sanitized[key.strip()] = value.strip()
        
    return sanitized

def validate_cwd(cwd: Optional[Path]) -> Optional[Path]:
    """
    Validate working directory
    
    Args:
        cwd: Working directory path
        
    Returns:
        Validated directory path
        
    Raises:
        CommandValidationError: If directory is invalid
    """
    if not cwd:
        return None
        
    if not isinstance(cwd, Path):
        cwd = Path(cwd)
        
    if not cwd.exists():
        raise CommandValidationError(f"Working directory does not exist: {cwd}")
    
    if not cwd.is_dir():
        raise CommandValidationError(f"Path is not a directory: {cwd}")
        
    return cwd

File: core/executer/src/utils/validation.rs
use std::collections::HashMap;
use std::path::PathBuf;
use crate::ExecuterError;

pub fn validate_command(cmd: &[String]) -> Result<(), ExecuterError> {
    if cmd.is_empty() {
        return Err(ExecuterError::ValidationError("Empty command sequence".to_string()));
    }

    for arg in cmd {
        if arg.contains(['&', '|', ';', '`', '\\']) {
            return Err(ExecuterError::ValidationError(
                format!("Invalid command argument: {}", arg)
            ));
        }
    }
    Ok(())
}

pub fn validate_env(env: &HashMap<String, String>) -> Result<(), ExecuterError> {
    for (k, v) in env {
        if k.trim().is_empty() || v.trim().is_empty() {
            return Err(ExecuterError::ValidationError(
                "Environment variables cannot be empty".to_string()
            ));
        }
    }
    Ok(())
}

pub fn validate_cwd(cwd: &Option<PathBuf>) -> Result<(), ExecuterError> {
    if let Some(path) = cwd {
        if !path.exists() {
            return Err(ExecuterError::ValidationError(
                format!("Working directory does not exist: {}", path.display())
            ));
        }
        if !path.is_dir() {
            return Err(ExecuterError::ValidationError(
                format!("Path is not a directory: {}", path.display())
            ));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_command() {
        assert!(validate_command(&["ls".to_string(), "-l".to_string()]).is_ok());
        assert!(validate_command(&[]).is_err());
        assert!(validate_command(&["ls".to_string(), "&".to_string()]).is_err());
    }

    #[test]
    fn test_validate_env() {
        let mut env = HashMap::new();
        env.insert("KEY".to_string(), "value".to_string());
        assert!(validate_env(&env).is_ok());

        env.insert("EMPTY".to_string(), "".to_string());
        assert!(validate_env(&env).is_err());
    }

    #[test]
    fn test_validate_cwd() {
        assert!(validate_cwd(&None).is_ok());
        assert!(validate_cwd(&Some(PathBuf::from("."))).is_ok());
        assert!(validate_cwd(&Some(PathBuf::from("/nonexistent"))).is_err());
    }
}